### PR TITLE
Groups

### PR DESCRIPTION
**Pull request by YaremaKryven**
Added models for Groups, Shedule, Invitation and forms with validator

### CODE REVIEW

File: groups/forms.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+class InvitationForm(forms.ModelForm):
+    """
+    Makes URL by which users can join to group
+    """
+
+    def clean(self):
+        """Checking validity of dates and number of students should be >0"""
+        cleaned_data = super(InvitationForm, self).clean()
+        errors = dates_validator(data={
```
Comment (MrKreg) [inline groups/forms.py]: It`ll be better to give as "data" parameter self.cleaned_data.

File: groups/forms.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+
+class InvitationForm(forms.ModelForm):
+    """
+    Makes URL by which users can join to group
+    """
+
+    def clean(self):
+        """Checking validity of dates and number of students should be >0"""
+        cleaned_data = super(InvitationForm, self).clean()
```
Comment (MrKreg) [inline groups/forms.py]: Django version of clean method sets it`s result to self.cleaned_data. 
Yout variable cleaned_data is useless.

**Review by MrKreg — CHANGES_REQUESTED** [review]

File: groups/validators.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+        if end < timezone.now():
+            msg = u'End date already passed! Please enter valid date.'
+            errors['end'] = msg
+        if not data['items']:
+            msg = u'Input number of students to invite to this group.'
+            errors['items_left'] = msg
+    if start >= end:
+        msg = u'End date should be after start date!'
+        errors['end'] = msg
+    return errors
```
Comment (rostIvan) [inline groups/validators.py]: Could it looks something like this?
``` python 
class ScheduleValidator:
    @staticmethod
    def validate(begin, end):
        correct_date = timezone.now() < begin < end
        if not correct_date:
            raise ValidationError('Date is incorrect')


class InvitationValidator:
    @staticmethod
    def validate(begin, end, slots_left):
        ScheduleValidator.validate(begin, end)
        if slots_left == 0:
            raise NoAvailableStudentSlotsError()


class NoAvailableStudentSlotsError(Exception):
    def __init__(self, *args: object, **kwargs: object) -> None:
        self.message = 'None available students slots left'
        super().__init__(self.message, *args, **kwargs)
```

File: groups/views.py [hunk 4]
```diff
@@ -1,3 +1,161 @@
...
+from django.db.models import Count
+from django.shortcuts import get_object_or_404
+from django.urls import reverse_lazy, reverse
+from django.views import generic
+
+from groups.forms import GroupCreateForm
+from groups.models import Group
+from CultureAnalyzer.mixins import SafePaginationMixin
+
+PAGINATOR = 50
```
Comment (falled10) [inline groups/views.py]: can you change name of this variable?

File: groups/views.py [hunk 5]
```diff
@@ -1,3 +1,161 @@
...
+        concatenates them with unchecked mentors, so that
+        checked mentors are always first in list
+        Search by last_name of mentor is available
+        """
+        checked_mentors = get_user_model().objects.filter(
+            is_active=True, mentor_in_group=self.kwargs['pk']). \
+            order_by('last_name')
+        self._checked_mentors = checked_mentors
+        mentors = get_user_model().objects. \
+            filter(is_active=True, groups__name='Mentor'). \
```
Comment (falled10) [inline groups/views.py]: Can you remove magic string 'Mentor'? You can use for example METOR_ID from CultureAnalyzer.constants

File: groups/validators.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
+from django.utils import timezone
+from CultureAnalyzer.exceptions import PValidationError
```
Comment (falled10) [inline groups/validators.py:3]: According to pep-8 a blank like should be added before imports from current project

**Review by falled10 — CHANGES_REQUESTED** [review]

File: groups/views.py (hunk 4 above)
Comment (YaremaKryven) [reply groups/views.py]: Changed to constants/ITEMS_ON_PAGE (imported from CultureAnalyser/constants)

File: groups/views.py (hunk 5 above)
Comment (YaremaKryven) [reply groups/views.py]: changed to groups__pk=MENTOR_ID (imported from CultureAnalyser/constants)

**Review by falled10 — APPROVED** [review]

File: groups/urls.py [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+app_name = 'groups'
+
+urlpatterns = [
+    path('', GroupsList.as_view(), name='groups-list'),
+    path('create-group/', CreateGroupView.as_view(), name='create-group'),
+    path('delete-group/<int:pk>', DeleteGroupView.as_view(),
+         name='delete-group'),
+    path('update-group/<int:pk>', UpdateGroupView.as_view(),
+         name='update-group'),
+
```
Comment (MrKreg) [inline groups/urls.py:14]: Can you remove this empty line?

**Review by MrKreg — APPROVED** [review]

**Review by kolinkovlad — APPROVED** [review]

**Pull request merged**